### ISSUE TITLE
Services table overflows mobile screen

### ISSUE THREAD

**Issue by garrett** [cockpit-project/cockpit#11260]
The lists we use are not accessible nor are they responsive. It's a big undertaking to fix either of these issues, but they're intertwingled so much that we'd have to address both at the same time.

**Comment by garrett** [comment]
After a good bit of research, the best examples of how to solve lists for desktops, mobile, and screen readers all at the same time seem to be to use a mix of techniques and HOWTOs at:

- [A11Y-101's Tables primer](https://www.a11y-101.com/development/tables)
- [A11Y Project "How-to: Create accessible data tables"](https://a11yproject.com/posts/accessible-data-tables/)
- [Charlie Cathcart's CodePen "top 10 grossing animated films" example](https://codepen.io/pixelchar/pen/rfuqK) utilizing `content: attr(data-title);` in a `td[data-title]:before`
- Adrian Roselli's posts
  - ["A Responsive Accessible Table" (part 1, 2017)](http://adrianroselli.com/2017/11/a-responsive-accessible-table.html)
    - The [Scrolling approach](http://adrianroselli.com/2017/11/a-responsive-accessible-table.html#ResponsiveScrolling) is interesting for some cases.
   - ["Tables, CSS Display Properties, and ARIA" (part 2, follow-up post from 2018)](http://adrianroselli.com/2018/02/tables-css-display-properties-and-aria.html)
    - [CodePen example](https://codepen.io/aardrian/pen/LQddwO/)

There are others out on the Internet that I had to wade through, but they had one issue or another. Or they claimed they were accessible but were doing things like changing the display without having fallbacks (they were lacking `scope="row"`/`scope="col"` in `th`s and ARIA... so they _couldn't_ have been truly accessible).

**Comment by garrett** [comment]
This issue (and the related research) was inspired by https://github.com/cockpit-project/cockpit/issues/11258.

I looked to see if we already had this issue, but we apparently didn't. We just all kind of knew in the back of our heads that the lists need fixing for responsiveness and accessibility. Anyway, this issue makes it formal. :wink:

We probably need *both* the scrolling approach and the deconstruction one too. I _think_ we should be able to use the same HTML for both methods and add a class to allow it to self-deconstruct in tight spaces. (In other words, add a variable that gets passed as a class for deconstruction. Or to prevent deconstruction... depending on what we want as a default.)

**Comment by garrett** [comment]
## Screenshots of various list issues on mobile

System logs truncates too agressively; information isn't spaced properly:

![system logs](https://user-images.githubusercontent.com/10246/56119630-cc49b500-5f6c-11e9-8db5-2cf446c37bef.jpg)

Services table runs off the page (currently being addressed in #11334, although perhaps the list part should be split out):
![thumb0033](https://user-images.githubusercontent.com/10246/56119703-ee433780-5f6c-11e9-9e0c-50a5c9ac8ab8.jpg)

Session recording list is just plainly odd:
![thumb0042](https://user-images.githubusercontent.com/10246/56119772-1468d780-5f6d-11e9-9f59-c375e035393d.jpg)

## Solutions

1. Remove unnecessary information (for information when the item is collapsed, as in the case of Software Updates)
2. Make table scroll off the side and be horizontally slidable
3. Deconstruct table in an accessible manner

Option 1 is the easiest, but works well only in a few small areas — in Software Updates, the changelog is teased and we simply drop that as it's duplicated in full in the expanded area. It's also not necessary when browsing the list.

Option 3 is the next best solution, as it displays all the same information on mobile and provides context.

Option 2 is tricky, as it needs to be clear that there is more information if someone slides to the side. It's still better than the mess of lists we have in some cases.

**Comment by martinpitt** [comment]
The scope of this is waaaaay to big to be tracked in a single issue. We've been working on porting our lists to PF4 page by page, slowly improving them as we go. So I suggest splitting that up into one issue per table, perhaps starting with the Services table. I retitle this issue accordingly -- while it looks a bit different now, it's still essentially the same problem: The "State" column gets cut off, and "Automatic startup" isn't visible at all.

**Comment by garrett** [comment]
In git commit 316711678772da3233fca57e6386dfd8dae3aab4, I fixed all of Cockpit's panels overflowing on mobile by letting panels be wide and scroll offscreen with a visual indicator, making it swipeable. For logs, I also added a minimum width so wrapping wouldn't be so horrible (opting to overflow for the swipe).

As a result, I consider this fixed. :+1:

**Comment by garrett** [comment]
Of course, when this is ported to PF4, it will be handled differently, but we'll handle that in the future. (It may just collapse in an optimal way then anyway. :man_shrugging: )

**Comment by martinpitt** [comment]
@garrett: I tried this on latest cockpit yesterday, so it's really not, I'm afraid :/

![service-mobile](https://user-images.githubusercontent.com/200109/67967548-1000c500-fc06-11e9-94e1-067d6b194081.png)

**Comment by garrett** [comment]
What's the problem? The table cannot be collapsed, so it flows off the page, and you swipe to reveal more. There's the shade at the end to indicate this. It's a typical pattern on mobile for both apps and websites (and not everything has an overflow indicator like we do even...)

Before, it just flowed without a way to see more information.

We cannot deconstruct the table for mobile and have it be accessible without _a lot_ of work, so this is the only thing we can do. PF4 may change this, as stated above.

As this needs to be rewritten anyway (as this isn't React nor is it even really PF3, but some custom jQuery stuff, IIRC), it doesn't make sense to spend all the time and effort to make the table deconstruct.

The issue with overflowing without being able to get to the information has been fixed in the commit I referenced above.

**Comment by garrett** [comment]
Of course, the problem with the list refreshing is still not fixed, so any more effort to make this work even better on mobile is completely wasted. 

It simply needs to be rewritten in React (so inline updates won't obliterate the page every few seconds) with PF4 before we invest even more time in it.

**Comment by martinpitt** [comment]
Ah, sorry for misunderstanding that then!
